### ISSUE TITLE
Freeze when trying to check a file exists

### ISSUE THREAD

**Issue by soroshsabz** [dsplaisted/PCLStorage#46]
ITNOA

Hi,

When I try to check exists file such as below code

``` C#
public static async Task<Stream> GetStreamAsync(string path)
{
	IFolder rootFolder = FileSystem.Current.LocalStorage;
	ExistenceCheckResult isExistFile = await rootFolder.CheckExistsAsync(path);
	if (isExistFile == ExistenceCheckResult.NotFound)
		throw new FileNotFoundException($"path: {path}");
	IFile file = await rootFolder.GetFileAsync(path);
	return await file.OpenAsync(FileAccess.Read);
}
```

My process freeze, I found one problem similar to me from [here](https://forums.xamarin.com/discussion/17642/freeze-when-trying-to-read-a-file-using-pclstorage).

**Comment by bgroenks96** [comment]
This appears to be a bug with the behavior of the method [`AwaitExtensions.SwitchOffMainThreadAsync`](https://github.com/dsplaisted/PCLStorage/blob/master/src/PCLStorage/AwaitExtensions.cs#L23).

The workaround (at least that I've found) is to spawn a new thread using `ThreadPool` or the `Thread` class to call the PCLStorage APIs. You can use a `TaskCompletionSource` to asynchronously await the results of the new thread's task.

**Comment by JKennedy24** [comment]
I am also experiencing this bug

**Comment by tmm360** [comment]
+1

**Comment by tmm360** [comment]
Can you fix this? I'm using a PCL library with profile44, I don't have `Thread` class, and importing `ThreadPool` requires `Windows.Foundation.UniversalApiContract` and `Windows.Foundation.FoundationContract` that gives problems to me.

**Comment by tmm360** [comment]
Also running on an other thread doesn't work for me. I've tried to implement this into the UWP solution, and inject it into the PCL library where I'm using PCLStorage. It still freeze on `await Task.Delay(10)`.

    class AsyncRunner : IAsyncRunner
    {
        public async Task RunOnOtherThread(Action action)
        {
            var asyncAction = ThreadPool.RunAsync(_ => action());
            while (asyncAction.Status != Windows.Foundation.AsyncStatus.Completed)
                await Task.Delay(10);
        }
    }

**Comment by soroshsabz** [comment]
Try some function like below

``` C#
public static async Task<ExistenceCheckResult> CheckExists(this IFolder rootFolder, string fileName)
{
	return await Task.Run(() => rootFolder.CheckExistsAsync(fileName)).ConfigureAwait(false);
}
```

and call it synchronous with get Result to resolve your problem temporary.

**Comment by tmm360** [comment]
Thank you, it works but only sometimes. It still can freeze, on CheckExists and on other calls. We really need a fix...

**Comment by soroshsabz** [comment]
Yes, you right. I need to fix it, but when I call all asynchronous function with top pattern, I could not see freeze any more. Did you sure call all API with top pattern?

**Comment by tmm360** [comment]
I've upgrade every single call to the library with your pattern, and it seems to work now. Before I had OpenAsync on IFile to update, and it froze on alternate moments. Thank you! Are you a maintainer?

**Comment by soroshsabz** [comment]
You are welcome :)
No, I am not maintainer. Unfortunately, I do not know why this library is not maintained continuously.

**Comment by thumbmunkeys** [comment]
I think the error is here:

`  
         public async Task<ExistenceCheckResult> CheckExistsAsync(string name, CancellationToken cancellationToken)
        {
             var result = await _wrappedFolder.GetItemAsync(name).AsTaskNoThrow(cancellationToken);`

ConfigureAwait(false) is missing

**Comment by JKennedy24** [comment]
From the comments this seems like a 1 line change, but as the last check in was Jan 2015. I'm not holding my breath. Frustrating

**Comment by iWeaverMan** [comment]
nuget update?

**Comment by mentat9** [comment]
+1
Getting this fix in the nuget package would be most helpful: still repros on 1.0.2. Not surprising since it was published 1/18/2015. Can someone make another build, so we can get the fixed library via nuget? Help with this would be most appreciated!

**Comment by JKennedy24** [comment]
fixed this in the source ages ago. Still no nuget update!

**Comment by Tamachan87** [comment]
2019, getting this problem with the Nuget package too. Can we update this package?

**Comment by JKennedy24** [comment]
@Tamachan87 Ditch this nuget entirely and swap it all for `File` and `Directory` which have now been implemented for .Net standard 2.0
